src/peft/tuners/tuners_utils.py

```diff
@@ -32,6 +35,56 @@
 logger = logging.getLogger(__name__)
 
 
+@contextmanager
+def onload_layer(layer):
+    r"""
+    A utility for modifying a module containing one or more tuners and a base layer, any of which are offloaded. Moves
+    a module's sub-modules to the execution device before some action is performed, after that that the state
+    dictionary is re-assigned and parameters are offloaded. If the module has no offloaded sub-modules, this function
+    does nothing.
+
+    Args:
+        layer ('torch.nn.Module'):
+            layer with tuners to be merged
+    """
+
+    offloaded_modules = []
+    for name, module in layer.named_modules():
+        if name in ["", "base_layer"]:
+            continue
+        if hasattr(module, "_hf_hook") and isinstance(module._hf_hook, AlignDevicesHook) and module._hf_hook.offload:
+            module._hf_hook.pre_forward(module)
+            offloaded_modules.append(module)
+
+    base_layer_offload = False
+    if hasattr(layer, "base_layer") and (
+        hasattr(layer.base_layer, "_hf_hook")
+        and isinstance(layer.base_layer._hf_hook, AlignDevicesHook)
+        and layer.base_layer._hf_hook.offload
+    ):
+        if torch.device("meta") in layer.base_layer._hf_hook.original_devices.values():
+            # retrieve the name of the original disk-offload directory
+            offload_folder = layer.base_layer._hf_hook.weights_map.dataset.save_folder
+        layer.base_layer._hf_hook.pre_forward(layer.base_layer)
+        base_layer_offload = True
+
+    yield
+
+    for module in offloaded_modules:
+        module._hf_hook.post_forward(module, torch.tensor([]))
+
+    if base_layer_offload:
+        # re-make weights map (must be on cpu to send params to the disk via memmap if disk offload)
+        layer.base_layer._hf_hook.weights_map = {
+            name: param.to("cpu") for name, param in named_module_tensors(layer.base_layer)
+        }
+        # offload weights map to disk if original device is the disk
+        if torch.device("meta") in layer.base_layer._hf_hook.original_devices.values():
+            # rewrite directory with merged weights
+            offload_state_dict(offload_folder, layer.base_layer._hf_hook.weights_map)
```

Comment:
It is true that the test only checks CPU offloading even though the context manager function is designed for both CPU and disk offloading. This is mostly because `PeftModel.from_pretrained` is currently incompatible with disk offloading (as far as I know) due the disk offload directory not being accessed from the base model properly.  I am happy to work on fixing this problem in the near future so the disk offloading test may be added, but thought it might be out-of-scope for this PR.  In the meantime I have added a todo comment for disk offloading in the test:)

The context manager and test description docstrings have been updated to reflect which offloading destinations each currently supports.

Reply:
Thanks for clarifying this ! While we are unable to test the disk offload, the code looks good. So, I'm not against keeping it.